File: frontend/src/utils/validation.js
/**
 * Validation utility functions
 * 
 * This module provides client-side validation functions for forms,
 * data validation, and input sanitization.
 * 
 * @author AI Job Readiness Team
 * @version 1.0.0
 */

import { VALIDATION_RULES, FILE_CONFIG } from './constants';

/**
 * Validate email format
 * @param {string} email - Email to validate
 * @returns {Object} - Validation result with isValid and message
 */
export const validateEmail = (email) => {
  if (!email) {
    return { isValid: false, message: 'Email is required' };
  }

  if (!VALIDATION_RULES.EMAIL.PATTERN.test(email)) {
    return { isValid: false, message: VALIDATION_RULES.EMAIL.MESSAGE };
  }

  return { isValid: true, message: '' };
};

/**
 * Validate password strength
 * @param {string} password - Password to validate
 * @returns {Object} - Validation result with isValid, message, and strength
 */
export const validatePassword = (password) => {
  if (!password) {
    return { isValid: false, message: 'Password is required', strength: 0 };
  }

  if (password.length < VALIDATION_RULES.PASSWORD.MIN_LENGTH) {
    return { 
      isValid: false, 
      message: `Password must be at least ${VALIDATION_RULES.PASSWORD.MIN_LENGTH} characters long`,
      strength: 0
    };
  }

  let strength = 0;
  const checks = {
    length: password.length >= VALIDATION_RULES.PASSWORD.MIN_LENGTH,
    uppercase: VALIDATION_RULES.PASSWORD.REQUIRE_UPPERCASE ? /[A-Z]/.test(password) : true,
    lowercase: VALIDATION_RULES.PASSWORD.REQUIRE_LOWERCASE ? /[a-z]/.test(password) : true,
    digits: VALIDATION_RULES.PASSWORD.REQUIRE_DIGITS ? /\d/.test(password) : true,
    special: VALIDATION_RULES.PASSWORD.REQUIRE_SPECIAL_CHARS ? /[!@#$%^&*(),.?":{}|<>]/.test(password) : true,
  };

  strength = Object.values(checks).filter(Boolean).length;

  if (!checks.uppercase || !checks.lowercase || !checks.digits) {
    return { 
      isValid: false, 
      message: VALIDATION_RULES.PASSWORD.MESSAGE,
      strength
    };
  }

  return { isValid: true, message: '', strength };
};

/**
 * Validate phone number format
 * @param {string} phone - Phone number to validate
 * @returns {Object} - Validation result with isValid and message
 */
export const validatePhone = (phone) => {
  if (!phone) {
    return { isValid: true, message: '' }; // Phone is optional
  }

  if (!VALIDATION_RULES.PHONE.PATTERN.test(phone)) {
    return { isValid: false, message: VALIDATION_RULES.PHONE.MESSAGE };
  }

  return { isValid: true, message: '' };
};

/**
 * Validate name format
 * @param {string} name - Name to validate
 * @param {string} fieldName - Field name for error message
 * @returns {Object} - Validation result with isValid and message
 */
export const validateName = (name, fieldName = 'Name') => {
  if (!name) {
    return { isValid: false, message: `${fieldName} is required` };
  }

  if (name.length < VALIDATION_RULES.NAME.MIN_LENGTH) {
    return { 
      isValid: false, 
      message: `${fieldName} must be at least ${VALIDATION_RULES.NAME.MIN_LENGTH} characters long` 
    };
  }

  if (name.length > VALIDATION_RULES.NAME.MAX_LENGTH) {
    return { 
      isValid: false, 
      message: `${fieldName} must be no more than ${VALIDATION_RULES.NAME.MAX_LENGTH} characters long` 
    };
  }

  if (!VALIDATION_RULES.NAME.PATTERN.test(name)) {
    return { isValid: false, message: VALIDATION_RULES.NAME.MESSAGE };
  }

  return { isValid: true, message: '' };
};

/**
 * Validate file upload
 * @param {File} file - File to validate
 * @returns {Object} - Validation result with isValid and message
 */
export const validateFile = (file) => {
  if (!file) {
    return { isValid: false, message: 'File is required' };
  }

  // Check file size
  if (file.size > FILE_CONFIG.MAX_SIZE) {
    return { 
      isValid: false, 
      message: `File size must be less than ${FILE_CONFIG.MAX_SIZE / (1024 * 1024)}MB` 
    };
  }

  // Check file type
  const fileExtension = '.' + file.name.split('.').pop().toLowerCase();
  if (!FILE_CONFIG.ALLOWED_TYPES.includes(fileExtension)) {
    return { 
      isValid: false, 
      message: `File type must be one of: ${FILE_CONFIG.ALLOWED_TYPES.join(', ')}` 
    };
  }

  // Check MIME type
  if (!FILE_CONFIG.ALLOWED_MIME_TYPES.includes(file.type)) {
    return { 
      isValid: false, 
      message: 'Invalid file type. Please upload a valid document.' 
    };
  }

  return { isValid: true, message: '' };
};

/**
 * Validate URL format
 * @param {string} url - URL to validate
 * @returns {Object} - Validation result with isValid and message
 */
export const validateUrl = (url) => {
  if (!url) {
    return { isValid: true, message: '' }; // URL is optional
  }

  try {
    new URL(url);
    return { isValid: true, message: '' };
  } catch {
    return { isValid: false, message: 'Please enter a valid URL' };
  }
};

/**
 * Validate required field
 * @param {any} value - Value to validate
 * @param {string} fieldName - Field name for error message
 * @returns {Object} - Validation result with isValid and message
 */
export const validateRequired = (value, fieldName) => {
  if (value === null || value === undefined || value === '') {
    return { isValid: false, message: `${fieldName} is required` };
  }

  return { isValid: true, message: '' };
};

/**
 * Validate minimum length
 * @param {string} value - Value to validate
 * @param {number} minLength - Minimum length required
 * @param {string} fieldName - Field name for error message
 * @returns {Object} - Validation result with isValid and message
 */
export const validateMinLength = (value, minLength, fieldName) => {
  if (!value || value.length < minLength) {
    return { 
      isValid: false, 
      message: `${fieldName} must be at least ${minLength} characters long` 
    };
  }

  return { isValid: true, message: '' };
};

/**
 * Validate maximum length
 * @param {string} value - Value to validate
 * @param {number} maxLength - Maximum length allowed
 * @param {string} fieldName - Field name for error message
 * @returns {Object} - Validation result with isValid and message
 */
export const validateMaxLength = (value, maxLength, fieldName) => {
  if (value && value.length > maxLength) {
    return { 
      isValid: false, 
      message: `${fieldName} must be no more than ${maxLength} characters long` 
    };
  }

  return { isValid: true, message: '' };
};

/**
 * Validate numeric value
 * @param {any} value - Value to validate
 * @param {string} fieldName - Field name for error message
 * @param {Object} options - Validation options
 * @returns {Object} - Validation result with isValid and message
 */
export const validateNumber = (value, fieldName, options = {}) => {
  const { min, max, integer = false } = options;

  if (value === null || value === undefined || value === '') {
    return { isValid: false, message: `${fieldName} is required` };
  }

  const num = Number(value);
  if (isNaN(num)) {
    return { isValid: false, message: `${fieldName} must be a valid number` };
  }

  if (integer && !Number.isInteger(num)) {
    return { isValid: false, message: `${fieldName} must be a whole number` };
  }

  if (min !== undefined && num < min) {
    return { isValid: false, message: `${fieldName} must be at least ${min}` };
  }

  if (max !== undefined && num > max) {
    return { isValid: false, message: `${fieldName} must be no more than ${max}` };
  }

  return { isValid: true, message: '' };
};

/**
 * Validate date
 * @param {any} value - Value to validate
 * @param {string} fieldName - Field name for error message
 * @param {Object} options - Validation options
 * @returns {Object} - Validation result with isValid and message
 */
export const validateDate = (value, fieldName, options = {}) => {
  const { min, max, future = false, past = false } = options;

  if (!value) {
    return { isValid: false, message: `${fieldName} is required` };
  }

  const date = new Date(value);
  if (isNaN(date.getTime())) {
    return { isValid: false, message: `${fieldName} must be a valid date` };
  }

  const now = new Date();
  if (future && date <= now) {
    return { isValid: false, message: `${fieldName} must be in the future` };
  }

  if (past && date >= now) {
    return { isValid: false, message: `${fieldName} must be in the past` };
  }

  if (min && date < new Date(min)) {
    return { isValid: false, message: `${fieldName} must be after ${min}` };
  }

  if (max && date > new Date(max)) {
    return { isValid: false, message: `${fieldName} must be before ${max}` };
  }

  return { isValid: true, message: '' };
};

/**
 * Validate form data
 * @param {Object} data - Form data to validate
 * @param {Object} rules - Validation rules
 * @returns {Object} - Validation result with isValid, errors, and data
 */
export const validateForm = (data, rules) => {
  const errors = {};
  let isValid = true;

  for (const [field, fieldRules] of Object.entries(rules)) {
    const value = data[field];
    const fieldError = validateField(value, fieldRules, field);
    
    if (!fieldError.isValid) {
      errors[field] = fieldError.message;
      isValid = false;
    }
  }

  return { isValid, errors, data };
};

/**
 * Validate individual field based on rules
 * @param {any} value - Field value
 * @param {Object} rules - Field validation rules
 * @param {string} fieldName - Field name
 * @returns {Object} - Validation result with isValid and message
 */
export const validateField = (value, rules, fieldName) => {
  // Required validation
  if (rules.required) {
    const requiredResult = validateRequired(value, fieldName);
    if (!requiredResult.isValid) {
      return requiredResult;
    }
  }

  // Skip other validations if field is empty and not required
  if (!value && !rules.required) {
    return { isValid: true, message: '' };
  }

  // Type-specific validations
  if (rules.type === 'email') {
    return validateEmail(value);
  }

  if (rules.type === 'password') {
    return validatePassword(value);
  }

  if (rules.type === 'phone') {
    return validatePhone(value);
  }

  if (rules.type === 'name') {
    return validateName(value, fieldName);
  }

  if (rules.type === 'url') {
    return validateUrl(value);
  }

  if (rules.type === 'number') {
    return validateNumber(value, fieldName, rules);
  }

  if (rules.type === 'date') {
    return validateDate(value, fieldName, rules);
  }

  // Length validations
  if (rules.minLength) {
    const minLengthResult = validateMinLength(value, rules.minLength, fieldName);
    if (!minLengthResult.isValid) {
      return minLengthResult;
    }
  }

  if (rules.maxLength) {
    const maxLengthResult = validateMaxLength(value, rules.maxLength, fieldName);
    if (!maxLengthResult.isValid) {
      return maxLengthResult;
    }
  }

  // Custom validation function
  if (rules.custom && typeof rules.custom === 'function') {
    return rules.custom(value, fieldName);
  }

  return { isValid: true, message: '' };
};

/**
 * Sanitize string input
 * @param {string} input - Input to sanitize
 * @returns {string} - Sanitized input
 */
export const sanitizeString = (input) => {
  if (typeof input !== 'string') {
    return '';
  }

  return input
    .trim()
    .replace(/[<>]/g, '') // Remove potential HTML tags
    .replace(/[&<>"']/g, (match) => {
      const escapeMap = {
        '&': '&amp;',
        '<': '&lt;',
        '>': '&gt;',
        '"': '&quot;',
        "'": '&#x27;',
      };
      return escapeMap[match];
    });
};

/**
 * Sanitize HTML input
 * @param {string} input - HTML input to sanitize
 * @returns {string} - Sanitized HTML
 */
export const sanitizeHtml = (input) => {
  if (typeof input !== 'string') {
    return '';
  }

  // Basic HTML sanitization - remove script tags and dangerous attributes
  return input
    .replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, '')
    .replace(/on\w+="[^"]*"/gi, '')
    .replace(/javascript:/gi, '');
};

/**
 * Validate and sanitize form data
 * @param {Object} data - Form data
 * @param {Object} rules - Validation rules
 * @returns {Object} - Sanitized and validated data
 */
export const validateAndSanitizeForm = (data, rules) => {
  const sanitizedData = {};
  const errors = {};
  let isValid = true;

  for (const [field, value] of Object.entries(data)) {
    // Sanitize string values
    if (typeof value === 'string') {
      sanitizedData[field] = sanitizeString(value);
    } else {
      sanitizedData[field] = value;
    }

    // Validate field if rules exist
    if (rules[field]) {
      const fieldError = validateField(sanitizedData[field], rules[field], field);
      if (!fieldError.isValid) {
        errors[field] = fieldError.message;
        isValid = false;
      }
    }
  }

  return { isValid, errors, data: sanitizedData };
};

File: backend/app/utils/validation.py
"""
Custom validation utilities for AI Job Readiness Platform

This module provides custom validation functions that are used
throughout the application for data validation and sanitization.

Author: AI Job Readiness Team
Version: 1.0.0
"""

import re
import uuid
from datetime import datetime, date
from typing import Any, Optional, List, Dict, Union, Tuple
from pydantic import ValidationError as PydanticValidationError

from app.core.config import settings


def validate_password_strength(password: str) -> Tuple[bool, List[str]]:
    """
    Validate password strength based on configured requirements.
    
    Args:
        password: Password to validate
        
    Returns:
        Tuple[bool, List[str]]: (is_valid, error_messages)
    """
    errors = []
    
    if not password:
        errors.append("Password is required")
        return False, errors
    
    # Check minimum length
    if len(password) < settings.security.min_password_length:
        errors.append(f"Password must be at least {settings.security.min_password_length} characters long")
    
    # Check for uppercase letters
    if settings.security.require_uppercase and not any(c.isupper() for c in password):
        errors.append("Password must contain at least one uppercase letter")
    
    # Check for lowercase letters
    if settings.security.require_lowercase and not any(c.islower() for c in password):
        errors.append("Password must contain at least one lowercase letter")
    
    # Check for digits
    if settings.security.require_digits and not any(c.isdigit() for c in password):
        errors.append("Password must contain at least one digit")
    
    # Check for special characters
    if settings.security.require_special_chars and not re.search(r'[!@#$%^&*(),.?":{}|<>]', password):
        errors.append("Password must contain at least one special character")
    
    return len(errors) == 0, errors


def validate_json_data(data: Any, schema: Optional[Dict[str, Any]] = None) -> Tuple[bool, Optional[str]]:
    """
    Validate JSON data structure.
    
    Args:
        data: Data to validate
        schema: Optional schema to validate against
        
    Returns:
        Tuple[bool, Optional[str]]: (is_valid, error_message)
    """
    try:
        # Basic JSON structure validation
        if not isinstance(data, (dict, list, str, int, float, bool, type(None))):
            return False, "Invalid JSON data type"
        
        # If schema is provided, perform basic validation
        if schema and isinstance(data, dict):
            required_fields = schema.get("required", [])
            for field in required_fields:
                if field not in data:
                    return False, f"Missing required field: {field}"
        
        return True, None
        
    except Exception as e:
        return False, f"JSON validation error: {str(e)}"


def validate_uuid(uuid_string: str) -> bool:
    """
    Validate UUID string format.
    
    Args:
        uuid_string: String to validate as UUID
        
    Returns:
        bool: True if valid UUID, False otherwise
    """
    if not uuid_string:
        return False
    
    try:
        uuid.UUID(uuid_string)
        return True
    except ValueError:
        return False


def validate_date_range(start_date: Union[str, datetime, date], 
                       end_date: Union[str, datetime, date]) -> Tuple[bool, Optional[str]]:
    """
    Validate that start_date is before end_date.
    
    Args:
        start_date: Start date
        end_date: End date
        
    Returns:
        Tuple[bool, Optional[str]]: (is_valid, error_message)
    """
    try:
        # Convert strings to datetime if needed
        if isinstance(start_date, str):
            start_date = datetime.fromisoformat(start_date.replace('Z', '+00:00'))
        elif isinstance(start_date, date) and not isinstance(start_date, datetime):
            start_date = datetime.combine(start_date, datetime.min.time())
        
        if isinstance(end_date, str):
            end_date = datetime.fromisoformat(end_date.replace('Z', '+00:00'))
        elif isinstance(end_date, date) and not isinstance(end_date, datetime):
            end_date = datetime.combine(end_date, datetime.min.time())
        
        if start_date >= end_date:
            return False, "Start date must be before end date"
        
        return True, None
        
    except (ValueError, TypeError) as e:
        return False, f"Invalid date format: {str(e)}"


def validate_file_upload(file_data: Dict[str, Any]) -> Tuple[bool, List[str]]:
    """
    Validate file upload data.
    
    Args:
        file_data: File upload data dictionary
        
    Returns:
        Tuple[bool, List[str]]: (is_valid, error_messages)
    """
    errors = []
    
    # Check required fields
    required_fields = ["filename", "content_type", "size"]
    for field in required_fields:
        if field not in file_data:
            errors.append(f"Missing required field: {field}")
    
    if errors:
        return False, errors
    
    # Validate filename
    if not file_data.get("filename"):
        errors.append("Filename cannot be empty")
    
    # Validate content type
    if not file_data.get("content_type"):
        errors.append("Content type cannot be empty")
    
    # Validate file size
    file_size = file_data.get("size", 0)
    if not isinstance(file_size, int) or file_size <= 0:
        errors.append("Invalid file size")
    elif file_size > settings.file.max_file_size:
        max_size_mb = settings.file.max_file_size // (1024 * 1024)
        errors.append(f"File size exceeds maximum limit of {max_size_mb}MB")
    
    return len(errors) == 0, errors


def validate_pagination_params(page: int, per_page: int, max_per_page: int = 100) -> Tuple[bool, List[str]]:
    """
    Validate pagination parameters.
    
    Args:
        page: Page number
        per_page: Items per page
        max_per_page: Maximum items per page allowed
        
    Returns:
        Tuple[bool, List[str]]: (is_valid, error_messages)
    """
    errors = []
    
    if not isinstance(page, int) or page < 1:
        errors.append("Page must be a positive integer")
    
    if not isinstance(per_page, int) or per_page < 1:
        errors.append("Per page must be a positive integer")
    elif per_page > max_per_page:
        errors.append(f"Per page cannot exceed {max_per_page}")
    
    return len(errors) == 0, errors


def validate_search_query(query: str, min_length: int = 2, max_length: int = 100) -> Tuple[bool, Optional[str]]:
    """
    Validate search query parameters.
    
    Args:
        query: Search query string
        min_length: Minimum query length
        max_length: Maximum query length
        
    Returns:
        Tuple[bool, Optional[str]]: (is_valid, error_message)
    """
    if not query:
        return True, None  # Empty query is allowed
    
    if len(query) < min_length:
        return False, f"Search query must be at least {min_length} characters long"
    
    if len(query) > max_length:
        return False, f"Search query cannot exceed {max_length} characters"
    
    # Check for potentially malicious patterns
    malicious_patterns = [
        r'<script.*?>.*?</script>',
        r'javascript:',
        r'data:',
        r'vbscript:',
    ]
    
    for pattern in malicious_patterns:
        if re.search(pattern, query, re.IGNORECASE):
            return False, "Search query contains potentially malicious content"
    
    return True, None


def validate_email_domain(email: str, allowed_domains: Optional[List[str]] = None) -> bool:
    """
    Validate email domain against allowed domains.
    
    Args:
        email: Email address to validate
        allowed_domains: List of allowed domains (if None, all domains allowed)
        
    Returns:
        bool: True if domain is allowed, False otherwise
    """
    if not email or not allowed_domains:
        return True
    
    try:
        domain = email.split('@')[1].lower()
        return domain in [d.lower() for d in allowed_domains]
    except IndexError:
        return False


def validate_phone_format(phone: str, country_code: str = "US") -> Tuple[bool, Optional[str]]:
    """
    Validate phone number format.
    
    Args:
        phone: Phone number to validate
        country_code: Country code for validation
        
    Returns:
        Tuple[bool, Optional[str]]: (is_valid, error_message)
    """
    if not phone:
        return True, None  # Empty phone is allowed
    
    # Basic phone number validation
    phone_pattern = r'^[\+]?[1-9][\d]{0,15}$'
    if not re.match(phone_pattern, phone.replace(' ', '').replace('-', '').replace('(', '').replace(')', '')):
        return False, "Invalid phone number format"
    
    return True, None


def validate_score_range(score: float, min_score: float = 0.0, max_score: float = 100.0) -> Tuple[bool, Optional[str]]:
    """
    Validate score is within valid range.
    
    Args:
        score: Score to validate
        min_score: Minimum valid score
        max_score: Maximum valid score
        
    Returns:
        Tuple[bool, Optional[str]]: (is_valid, error_message)
    """
    if not isinstance(score, (int, float)):
        return False, "Score must be a number"
    
    if score < min_score or score > max_score:
        return False, f"Score must be between {min_score} and {max_score}"
    
    return True, None


def validate_analysis_type(analysis_type: str) -> Tuple[bool, Optional[str]]:
    """
    Validate analysis type is one of the allowed types.
    
    Args:
        analysis_type: Analysis type to validate
        
    Returns:
        Tuple[bool, Optional[str]]: (is_valid, error_message)
    """
    allowed_types = ["overall", "job_match", "skill_analysis", "experience_analysis", "education_analysis"]
    
    if analysis_type not in allowed_types:
        return False, f"Invalid analysis type. Must be one of: {', '.join(allowed_types)}"
    
    return True, None


def sanitize_input(text: str, max_length: Optional[int] = None) -> str:
    """
    Sanitize user input by removing potentially harmful content.
    
    Args:
        text: Text to sanitize
        max_length: Maximum length to truncate to
        
    Returns:
        str: Sanitized text
    """
    if not text:
        return ""
    
    # Remove HTML tags
    text = re.sub(r'<[^>]+>', '', text)
    
    # Remove script tags and javascript
    text = re.sub(r'<script.*?>.*?</script>', '', text, flags=re.IGNORECASE | re.DOTALL)
    text = re.sub(r'javascript:', '', text, flags=re.IGNORECASE)
    
    # Remove SQL injection patterns
    sql_patterns = [
        r'(\b(union|select|insert|update|delete|drop|create|alter)\b)',
        r'(\b(or|and)\s+\d+\s*=\s*\d+)',
        r'(\'|\"|;|--|\/\*|\*\/)',
    ]
    
    for pattern in sql_patterns:
        text = re.sub(pattern, '', text, flags=re.IGNORECASE)
    
    # Remove extra whitespace
    text = re.sub(r'\s+', ' ', text.strip())
    
    # Truncate if max_length is specified
    if max_length and len(text) > max_length:
        text = text[:max_length].rstrip()
    
    return text
